cmd/shadowsocks-local/local.go

```diff
@@ -1,430 +1,814 @@
+		return -1, err
+	}
+	return int(bufrq[idCmd]), nil
 }
 
-func getRequest(conn net.Conn) (rawaddr []byte, host string, err error) {
-	const (
-		idVer   = 0
-		idCmd   = 1
-		idType  = 3 // address type index
-		idIP0   = 4 // ip addres start index
-		idDmLen = 4 // domain address length index
-		idDm0   = 5 // domain address start index
-
-		typeIPv4 = 1 // type is ipv4 address
-		typeDm   = 3 // type is domain address
-		typeIPv6 = 4 // type is ipv6 address
-
-		lenIPv4   = 3 + 1 + net.IPv4len + 2 // 3(ver+cmd+rsv) + 1addrType + ipv4 + 2port
-		lenIPv6   = 3 + 1 + net.IPv6len + 2 // 3(ver+cmd+rsv) + 1addrType + ipv6 + 2port
-		lenDmBase = 3 + 1 + 1 + 2           // 3 + 1addrType + 1addrLen + 2port, plus addrLen
-	)
+// getRequest return the socks5 request from the client
+// return the type+addr+port as rawaddr as the request payload
+func getRequest(conn net.Conn) (rawaddr []byte, err error) {
+	var n int
 	// refer to getRequest in server.go for why set buffer size to 263
 	buf := make([]byte, 263)
-	var n int
-	ss.SetReadTimeout(conn)
-	// read till we get possible domain length field
-	if n, err = io.ReadAtLeast(conn, buf, idDmLen+1); err != nil {
-		return
-	}
-	// check version and cmd
-	if buf[idVer] != socksVer5 {
-		err = errVer
-		return
-	}
-	if buf[idCmd] != socksCmdConnect {
-		err = errCmd
-		return
+
+	// [ver* cmd* resv* type dmlen]
+	// first of all we read the socks5 request
+	// read till we get possible domain length
+	if n, err = io.ReadAtLeast(conn, buf[3:], idDm0); err != nil {
+		if err != io.EOF {
+			return
+		}
+		if n == 0 {
+			err = ErrReqExtraData
+			return
+		}
 	}
 
+	// cause we have read the first 3 byte (ver cmd reserve), the offset shoule be add here
+	// first 3 byte is meaningless
+	n += 3
+
 	reqLen := -1
-	switch buf[idType] {
+	switch buf[idType] & 0xff {
 	case typeIPv4:
 		reqLen = lenIPv4
 	case typeIPv6:
 		reqLen = lenIPv6
 	case typeDm:
 		reqLen = int(buf[idDmLen]) + lenDmBase
 	default:
-		err = errAddrType
+		err = ErrAddrType
 		return
 	}
 
 	if n == reqLen {
-		// common case, do nothing
+		// common case cause we have read all the info (host + port), do nothing
 	} else if n < reqLen { // rare case
+		ss.Logger.Warn("error in read the request arddr, less than expect", zap.Int("recv", n), zap.Int("req", reqLen))
 		if _, err = io.ReadFull(conn, buf[n:reqLen]); err != nil {
-			return
+			if err != io.EOF {
+				return
+			}
+			if n == 0 {
+				err = ErrReqExtraData
+				return
+			}
 		}
 	} else {
-		err = errReqExtraData
+		err = ErrReqExtraData
 		return
 	}
 
-	rawaddr = buf[idType:reqLen]
+	// only contain the type + host + port
+	rawaddr = buf[idType:n]
 
-	if debug {
-		switch buf[idType] {
-		case typeIPv4:
-			host = net.IP(buf[idIP0 : idIP0+net.IPv4len]).String()
-		case typeIPv6:
-			host = net.IP(buf[idIP0 : idIP0+net.IPv6len]).String()
-		case typeDm:
-			host = string(buf[idDm0 : idDm0+buf[idDmLen]])
+	var host string
```

Comment:
only compute the host when logger debug mode is enable

Reply:
nope, you only need to log out the log with corrcet level, and open the log